pkg/skaffold/log/provider.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package log
+
+import (
+	"sync"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/logger"
+)
+
+type Provider interface {
+	Do() Logger
+}
+
+type FullProvider struct {
+	tail bool
+
+	kubernetesLogger *logger.LogAggregator
+	noopLogger       *NoopLogger
+}
+
+var provider *FullProvider
+
+func NewLogProvider(cli *kubectl.CLI, podSelector kubernetes.PodSelector, config logger.Config) Provider {
+	once := sync.Once{}
+	once.Do(func() {
+		kLog := logger.NewLogAggregator(cli, podSelector, config)
+		provider = &FullProvider{
+			tail:             config.Tail(),
+			kubernetesLogger: kLog,
+			noopLogger:       &NoopLogger{},
+		}
+	})
+	return provider
```

Comment:
not sure what the intent is here.
Right now the `once` var is redeclared on every call to `NewLogProvider` so `klog` is not really a singleton.

Reply:
`once` and `provider` need to be declared outside the function.